### PR TITLE
RHCLOUD-45010: Add test coverage before RelationsRepository refactor

### PR DESCRIPTION
**Pull request by snehagunta**
## Summary

Build comprehensive test safety net using SimpleRelationsRepository (fake) before refactoring RelationsRepository interface.

**This PR contains test changes only** - no production code changes except stripping MockRelationsRepository to enforce "Fakes Over Mocks" pattern.

### 1. Config Tests (NEW)
- **File:** `internal/config/relations/config_test.go`
- 3 tests documenting current config completion behavior
- Documents bug where Kessel errors are swallowed

### 2. Migrated Tests  
- **File:** `internal/service/resources/kesselinventoryservice_test.go`
- Migrated 5 tests from MockRelationsRepository to SimpleRelationsRepository
- Tests: CheckSelf (allowed/denied), CheckSelfBulk (allowed/mixed)

### 3. Stripped MockRelationsRepository
- **File:** `internal/mocks/mocks.go`
- Removed 11 methods, kept only CheckBulk and CheckForUpdateBulk  
- Added panic stubs for removed methods
- Only 3 edge-case tests still use mock

### 4. Biz Layer Tests (NEW)
- **File:** `internal/biz/usecase/resources/resource_service_test.go`
- Added 9 tests: Check, CheckForUpdate, CheckBulk, LookupResources, LookupSubjects

### 5. API Layer Tests (NEW)  
- **File:** `internal/service/resources/kesselinventoryservice_test.go`
- Added 3 tests: StreamedListSubjects, empty stream tests

## Test Pattern: Fakes Over Mocks

All tests use SimpleRelationsRepository.Grant() pattern:
```go
simpleAuthz := data.NewSimpleRelationsRepository()
simpleAuthz.Grant("user-1", "view", "hbi", "host", "host-1")
```

## Test Summary
- Config tests: 3 new
- Biz layer: +9 tests  
- API layer: 5 migrated + 3 new
- **Total: 20 tests added/improved**

All tests passing ✅

## Next Steps
Follow-up PR will refactor RelationsRepository interface from Relations API protobuf types to Inventory domain types. The test coverage here ensures that refactor will be safe and regression-free.

## Related
- Jira: RHCLOUD-45010
- Implementation plan attached to PR and Jira ticket

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Broadened authorization coverage with table-driven allowed/denied cases, consolidated authz decision tests, deterministic streaming assertions, and a shared test harness to simplify per-test setup.
  * Replaced many heavy mocks with an in-memory relations helper and added config tests documenting implementation-selection behavior.

* **Chores**
  * Reduced mock surface to a minimal stub for select operations.
  * Enhanced the in-memory relations helper with injectable errors, lock-token behavior, and realistic consistency-token responses for testing.

### CODE REVIEW

**Comment by snehagunta** [comment]
---
name: Relations repository test coverage
overview: "Migrate mock-based tests to SimpleRelationsRepository, strip MockRelationsRepository of non-essential behavior, and add missing biz/API tests to make changing the RelationsRepository interface safe and regression-free."
todos:
  - id: config-tests
    content: "Add internal/config/relations/config_test.go (document current behavior, no production code changes)"
    status: pending
  - id: api-mock-migration
    content: "Migrate 5 API-layer mock-based tests to use SimpleRelationsRepository"
    status: pending
  - id: mock-cleanup
    content: "Remove all methods from MockRelationsRepository except CheckBulk and CheckForUpdateBulk (needed for error/mismatch edge cases)"
    status: pending
  - id: biz-deny-tests
    content: Add relations-deny and lookup streaming tests to resource_service_test.go
    status: pending
  - id: api-new-tests
    content: Add missing API-layer tests for bulk mixed results, lookup streaming, and empty streams
    status: pending
isProject: false
---

# Relations Repository Test Coverage Plan

## Goal

The `RelationsRepository` interface will change from using Relations API protobuf types to Inventory model types. This plan:
1. **Migrates** existing mock-based tests to `SimpleRelationsRepository` so they survive the interface change
2. **Strips** `MockRelationsRepository` down to only what can't be faked
3. **Adds** missing tests at biz and API layers to lock down observable behavior before the refactor

## Convention: Fakes Over Mocks

All tests should use `SimpleRelationsRepository` with `Grant()` for tuple setup. `MockRelationsRepository` is only kept for edge cases that `SimpleRelationsRepository` cannot produce (response length mismatch, per-pair errors).

---

## 1. Config Tests (No Production Code Changes)

**File:** `internal/config/relations/config_test.go` (new)

`config.go` line 42 has a known bug where `Complete` swallows Kessel errors. This plan documents the current behavior only; the fix is a separate change.

- `TestConfig_Complete_AllowAll_Success` — returns valid `CompletedConfig` with `Authz == AllowAll`
- `TestConfig_Complete_Kessel_SwallowsErrors` — kessel config that fails `Complete`; assert errors are currently `nil` (documents the bug)
- `TestCheckRelationsImpl_AllCases` — AllowAll -> "AllowAll", Kessel -> "Kessel", unknown -> "Unknown"

---

## 2. Migrate Mock-Based API Tests to SimpleRelationsRepository

**File:** [internal/service/resources/kesselinventoryservice_test.go](internal/service/resources/kesselinventoryservice_test.go)

5 tests currently use `MockRelationsRepository` but can be rewritten with `SimpleRelationsRepository`:

| Test (line) | Current mock setup | Replacement |
|---|---|---|
| `CheckSelf_Allowed_XRhIdentity` (491) | Mock `.On("Check", ...).Return(ALLOWED_TRUE)` with `mock.MatchedBy` on subject | `simpleAuthz.Grant("user-123", "view", "hbi", "host", "dd1b73b9-...")` |
| `CheckSelf_Allowed_XRhIdentity_SubjectIdMatch` (540) | Same pattern, different user id | `simpleAuthz.Grant("testuser", "view", "hbi", "host", "dd1b73b9-...")` |
| `CheckSelf_Denied` (586) | Mock returns `ALLOWED_FALSE` | Default `SimpleRelationsRepository` with no grants (same as `Check_Denied`) |
| `CheckSelfBulk_Allowed_XRhIdentity` (662) | Mock `CheckBulk` returns all allowed + asserts subject shape | `simpleAuthz.Grant(...)` for both items |
| `CheckSelfBulk_MixedResults` (759) | Mock `CheckBulk` returns one allowed, one denied | `simpleAuthz.Grant(...)` for resource-1/view only |

For each test:
- Replace `mockRelations := &mocks.MockRelationsRepository{}` with `simpleAuthz := data.NewSimpleRelationsRepository()` + `Grant()` calls
- Replace `testUsecaseConfig{Relations: mockRelations}` with `testUsecaseConfig{Relations: simpleAuthz}`
- Remove `mockRelations.AssertExpectations(t)`
- Keep all response assertions unchanged

Note: `CheckSelf_Allowed_XRhIdentity` currently asserts via `mock.MatchedBy` that the subject is correctly derived from claims (user id, type, namespace). With `SimpleRelationsRepository`, this is tested implicitly — if the subject derivation is wrong, the grant won't match and the test fails with `ALLOWED_FALSE`. The subject derivation logic itself is already tested in the `testSelfSubjectStrategy` / `buildTestSubjectReference` helpers.

---

## 3. Strip MockRelationsRepository

**File:** [internal/mocks/mocks.go](internal/mocks/mocks.go)

After migrating the 5 tests above, only 3 tests still need `MockRelationsRepository`:

| Test | Why mock is needed |
|---|---|
| `CheckSelfBulk_ResponseLengthMismatch` (815) | Returns more response pairs than request items — `SimpleRelationsRepository` always returns the correct count |
| `CheckForUpdateBulk_AllAllowed` (2862) | Uses `mock.MatchedBy` to verify subjects are correctly injected into the bulk request — can be migrated but is medium effort |
| `CheckForUpdateBulk_PairError` (3189) | Returns a per-pair `Status` error — `SimpleRelationsRepository` never produces per-pair errors |

Remove from `MockRelationsRepository` all methods **except**:
- `CheckBulk` — needed by `ResponseLengthMismatch` and potentially `CheckForUpdateBulk_AllAllowed`
- `CheckForUpdateBulk` — needed by `PairError`

Methods to remove:
- `Health`
- `Check`
- `CheckForUpdate`
- `LookupResources`
- `LookupSubjects`
- `CreateTuples`
- `DeleteTuples`
- `AcquireLock`
- `UnsetWorkspace`
- `SetWorkspace`

This will cause a compile error if `MockRelationsRepository` is used as a `model.RelationsRepository` directly. The 3 remaining tests that need it should embed it alongside a `SimpleRelationsRepository` for the non-mocked methods, or `MockRelationsRepository` should stop implementing the full interface and instead be injected via a wrapper. Alternatively, if these 3 tests can tolerate it, keep `MockRelationsRepository` implementing the full interface but with the unused methods panicking with a clear message.

---

## 4. Biz Layer — Add Missing Tests

**File:** [internal/biz/usecase/resources/resource_service_test.go](internal/biz/usecase/resources/resource_service_test.go) (add to existing)

### Current state
- 36 tests use `AllowAllRelationsRepository` — relations never denies
- 1 test uses `SimpleRelationsRepository` (`TestCheckForUpdateBulk_MixedResults`)
- 0 tests exercise `LookupResources` / `LookupSubjects` end-to-end through the usecase

### Tests to add

All use `SimpleRelationsRepository` with `Grant()` and permissive meta-authorizer:

- `TestCheck_RelationsDenied` — no tuples, meta allows, verify `ALLOWED_FALSE`
- `TestCheck_RelationsAllowed` — grant tuple, verify `ALLOWED_TRUE` + non-empty consistency token
- `TestCheckForUpdate_RelationsDenied` — no tuples, verify denied
- `TestCheckForUpdate_RelationsAllowed` — grant tuple, verify allowed
- `TestCheckBulk_RelationsMixedResults` — grant for some items, verify per-item results
- `TestLookupResources_EndToEnd` — grant tuples for 2 resources, consume stream, verify both IDs returned
- `TestLookupResources_Empty` — no tuples, stream returns EOF immediately
- `TestLookupSubjects_EndToEnd` — grant tuples for 2 subjects, verify both returned
- `TestLookupSubjects_Empty` — no tuples, verify EOF

---

## 5. API/Service Layer — Add Missing Tests

**File:** [internal/service/resources/kesselinventoryservice_test.go](internal/service/resources/kesselinventoryservice_test.go) (add to existing)

### Current state
- 62 tests use `SimpleRelationsRepository` (default or explicit)
- Good coverage for Check allowed/denied, CheckForUpdate, StreamedListObjects success
- Missing: bulk mixed results with `SimpleRelationsRepository`, list subjects, empty streams

### Tests to add

All use `SimpleRelationsRepository` with `Grant()`:

- `TestInventoryService_CheckBulk_RelationsMixedResults` — grant for some items, verify per-item allowed/denied via gRPC and HTTP
- `TestInventoryService_StreamedListSubjects_Success` — grant tuples, verify subjects returned via stream
- `TestInventoryService_StreamedListObjects_Empty` — no tuples, verify stream returns EOF cleanly
- `TestInventoryService_StreamedListSubjects_Empty` — no tuples, verify EOF

File: internal/biz/usecase/resources/resource_service_test.go [hunk 1]
```diff
@@ -23,6 +25,97 @@ import (
...
+	for _, o := range opts {
+		o(cfg)
+	}
+
+	resourceRepo := data.NewFakeResourceRepository()
+	mc := metricscollector.NewFakeMetricsCollector()
+	var metaAuth metaauthorizer.MetaAuthorizer
+	if cfg.meta != nil {
+		metaAuth = cfg.meta
+	}
```
[the remark this answers is not in the document]
Comment (Rajagopalan-Ranganathan) [reply internal/biz/usecase/resources/resource_service_test.go:64]: The current design is intentional and correct: withMeta(true/false) injects a recordingMetaAuthorizer when tests need to observe/control meta-auth calls; omitting it uses the production default via the constructor's nil-guard.

File: internal/data/relations_simple.go [hunk 2]
```diff
@@ -38,11 +38,21 @@ type simpleTupleKey struct {
...
+	mu                sync.RWMutex
+	version           int64                             // current version (monotonically increasing)
+	tuples            map[simpleTupleKey]bool           // current/latest tuple state
+	snapshots         map[int64]map[simpleTupleKey]bool // retained historical snapshots (version -> tuples)
+	healthError       error                             // if set, Health() returns this error
+	createTuplesError error                             // if set, CreateTuples() returns this error
+	deleteTuplesError error                             // if set, DeleteTuples() returns this error
+	acquireLockError  error                             // if set, AcquireLock() returns this error
+	locks             map[string]string                 // lockId -> token
 }
```
[the remark this answers is not in the document]
Comment (Rajagopalan-Ranganathan) [reply internal/data/relations_simple.go:56]: fixed

File: internal/config/relations/config_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	// Current behavior: when kessel.Complete succeeds, no errors
+	assert.Nil(t, errs, "When kessel.Complete succeeds, no errors should be returned")
+	assert.NotNil(t, completed.completedConfig, "Completed config should not be nil")
+	assert.Equal(t, Kessel, completed.Authz, "Authz should be Kessel")
+	assert.NotNil(t, completed.Kessel, "Kessel config should be present")
+
+	// The bug exists in the code but is documented here in comments:
+	// IF kessel.Complete were to return errors, they would be swallowed
+	// and nil would be returned instead (see config.go:42).
+}
```
[the remark this answers is not in the document]
Comment (Rajagopalan-Ranganathan) [reply internal/config/relations/config_test.go:69]: this is just a detailed comment explaining the beahviour. This shouldnt be marked as Major. Why is this needs fixing?
